src/Database.php

```diff
@@ -300,6 +304,59 @@ public function createCollection(string $collectionName, array $options = [])
         return $result;
     }
 
+    /**
+     * Create a new encrypted collection explicitly.
+     *
+     * This function will automatically create data keys for any encrypted
+     * fields where the "keyId" option is null. This function will return a copy
+     * of the modified "encryptedFields" option in addition to the result from
+     * createCollection(). The "encryptedFields" option is required.
+     *
+     * If any error is encountered while creating data keys or invoking
+     * createCollection(), a CreateEncryptedCollectionException will be thrown.
+     * The original exception and modified "encryptedFields" option can be
+     * accessed via the getPrevious() and getEncryptedFields() methods,
+     * respectively.
+     *
+     * @see CreateCollection::__construct() for supported options
+     * @return array A tuple consisting of the createCollection() result and modified "encryptedFields" option
+     * @throws InvalidArgumentException for parameter/option parsing errors
+     * @throws CreateEncryptedCollectionException for any errors creating data keys or invoking createCollection()
+     */
+    public function createEncryptedCollection(string $collectionName, ClientEncryption $clientEncryption, string $kmsProvider, ?array $masterKey, array $options): array
+    {
+        if (! isset($options['encryptedFields']) || ! is_array($options['encryptedFields']) && ! is_object($options['encryptedFields'])) {
+            throw InvalidArgumentException::invalidType('"encryptedFields" option', $options['encryptedFields'] ?? null, ['array', 'object']);
+        }
+
+        $encryptedFields = (array) recursive_copy($options['encryptedFields']);
```

Comment:
The following type would cover the `fields` property of the `encryptedFields` option:

```suggestion
        /** @var array{fields: list<array{path: string, bsonType: string, keyId: ?Binary}|object{path: string, bsonType: string, keyId: ?Binary}>} */
        $encryptedFields = (array) recursive_copy($options['encryptedFields']);
```

This could potentially be simplified by extracting the array shape and object shapes to separate [`@psalm-type`](https://psalm.dev/docs/annotating_code/supported_annotations/#psalm-type) annotations and referencing them here, in which case I'd suggest also using `@psalm-var` instead of `@var`.

To make this available to end users, we can also include the type in the type annotation for the method parameter as well. However, I'd rather wait until we use psalm v5 for this, as it changes the way array and list shapes are handled: https://psalm.dev/docs/annotating_code/type_syntax/array_types/#unsealed-array-and-list-shapes

Reply:
The structure for `encryptedFields` is defined in [CSFLE: Terms](https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#terms). It includes a bit more than what you have in the `@var` annotation, but I'm curious if that's even required. Should we only tell psalm about types to the extent that we're accessing data?

If so, a more concise form would be:

```
/** @var array{fields: list<array{keyId: ?Binary}|object{keyId: ?Binary}>} */
```

I'm also curious how `@psalm-type` works. The docs show it defined in a class-level doc block. Can it also be defined in a method doc block and apply to a more limited scope? Is there some higher scope (beyond a class) where you can define a type? I ask because this type isn't really specific to the Database class or this method, since it's shared between two helper methods and might also apply within the CreateCollection operation class.

A separate idea I wanted to explore in this PR (apart from revising the prose test organization) was to move the create/drop helpers with their CSFLE logic into separate operation classes. I'm not entirely satisfied with the current implementation, as validation of CreateCollection's options are all deferred until _after_ we generate data keys. I think there's room for improvement there, which would allow us to avoid needing to wrap InvalidArgumentExceptions. Also, refactoring the helpers into operations would reduce duplication between the drop helpers.

Reply:
> I'm also curious how `@psalm-type` works. The docs show it defined in a class-level doc block. Can it also be defined in a method doc block and apply to a more limited scope?

I'm not sure if they can be defined on a level lower than the class, I'd have to test that. As for a different scope, psalm offers the [`@psalm-import-type` annotation](https://psalm.dev/docs/annotating_code/supported_annotations/#psalm-import-type) to allow importing types being declared somewhere else. Since the annotation syntax seems to reference the class name where the type was defined on, I believe we may want to declare types only on the class level to allow importing them somewhere else if necessary.

To bring an example, let's say you were to extract this logic into an operation class, in which case you'd want to declare the type in the operation class as that's where you're handling it. We could then import the type using `@psalm-import-type` here and reference it again as if it were declared in the class.

Reply:
Operation refactoring was completed. I'll resolve this.